crates/matrix-sdk-ui/tests/integration/timeline/mod.rs

```diff
@@ -510,235 +510,166 @@ async fn test_duplicate_maintains_correct_order() {
-    let (client, server) = logged_in_client_with_server().await;
-    let sync_settings = SyncSettings::new().timeout(Duration::from_millis(3000));
-
-    let f = EventFactory::new().sender(user_id!("@a:b.c"));
-    let mut sync_builder = SyncResponseBuilder::new();
-    let event_id = event_id!("$a");
-    sync_builder.add_joined_room(JoinedRoomBuilder::new(room_id));
-
-    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
-    let _response = client.sync_once(sync_settings.clone()).await.unwrap();
-    server.reset().await;
-
-    let room = client.get_room(room_id).unwrap();
-    let timeline = room.timeline().await.unwrap();
-
-    sync_builder.add_joined_room(
-        JoinedRoomBuilder::new(room_id)
-            .add_timeline_event(f.text_msg("A").event_id(event_id).into_raw_sync())
-            .add_state_event(StateTestEvent::RoomPinnedEvents),
-    );
-    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
-    let _response = client.sync_once(sync_settings.clone()).await.unwrap();
+    let mut setup = PinningTestSetup::new().await;
+    let timeline = setup.timeline().await;
 
+    setup.mock_sync(true).await;
     assert!(!timeline.items().await.is_empty());
 
     // Unpinning a remote event fails.
-    Mock::given(method("PUT"))
-        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/state/m.room.pinned_events/.*?"))
-        .and(header("authorization", "Bearer 1234"))
-        .respond_with(ResponseTemplate::new(400))
-        .mount(&server)
-        .await;
+    setup.mock_response(ResponseTemplate::new(400)).await;
 
+    let event_id = setup.event_id();
     assert!(timeline.unpin_event(event_id).await.is_err());
 
-    server.reset().await;
+    setup.reset_server().await;
+}
+
+struct PinningTestSetup<'a> {
+    event_id: &'a ruma::EventId,
+    room_id: &'a ruma::RoomId,
+    client: matrix_sdk::Client,
+    server: wiremock::MockServer,
+    sync_settings: SyncSettings,
+    sync_builder: SyncResponseBuilder,
+}
+
+impl PinningTestSetup<'_> {
+    async fn new() -> Self {
+        let room_id = room_id!("!a98sd12bjh:example.org");
+        let (client, server) = logged_in_client_with_server().await;
+        let sync_settings = SyncSettings::new().timeout(Duration::from_millis(3000));
+
+        let mut sync_builder = SyncResponseBuilder::new();
+        let event_id = event_id!("$a");
+        sync_builder.add_joined_room(JoinedRoomBuilder::new(room_id));
+
+        mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+        let _response = client.sync_once(sync_settings.clone()).await.unwrap();
+        server.reset().await;
+
+        Self { event_id, room_id, client, server, sync_settings, sync_builder }
+    }
+
+    async fn timeline(&self) -> matrix_sdk_ui::Timeline {
+        let room = self.client.get_room(self.room_id).unwrap();
+        room.timeline().await.unwrap()
+    }
+
+    async fn reset_server(&self) {
```

Comment:
This could go inside an `impl Drop for PinningTestSetup` (I think), which would mean you did not have to remember to call it in the test.

Reply:
That would be cool, but drop is not async sadly

Reply:
Ha, of course!